File: modules/shared/src/main/scala/shared/topics/TopicNamer.scala
package shared.topics

import io.simplesource.kafka.api.ResourceNamingStrategy
import io.simplesource.kafka.util.PrefixResourceNamingStrategy

trait TopicNamer {
  def apply(topicType: String): String
}

object TopicNamer {
  def forStrategy(strategy: ResourceNamingStrategy,
                  topicBaseName: String,
                  allTopics: List[String]): TopicNamer =
    topicType => strategy.topicName(topicBaseName, topicType)

  def forPrefix(topicPrefix: String, baseName: String): TopicNamer =
    topicType => new PrefixResourceNamingStrategy(topicPrefix).topicName(baseName, topicType)
}

File: modules/shared/src/main/java/io/simplesource/shared/topics/TopicNamer.java
package io.simplesource.shared.topics;

import io.simplesource.kafka.api.ResourceNamingStrategy;
import io.simplesource.kafka.util.PrefixResourceNamingStrategy;

import java.util.List;

interface TopicNamer {
    String apply(String topicType);

    static TopicNamer forStrategy(ResourceNamingStrategy strategy,
                    String topicBaseName,
                    List<String> allTopics) {
        return topicType -> strategy.topicName(topicBaseName, topicType);
    }

    static TopicNamer forPrefix(String topicPrefix, String baseName) {
        return topicType -> new PrefixResourceNamingStrategy(topicPrefix).topicName(baseName, topicType);
    }
}
